README.md
# Introduction

[![Latest Version](https://img.shields.io/packagist/v/oanhnn/laravel-logzio.svg)](https://packagist.org/packages/oanhnn/laravel-logzio)
[![Software License](https://img.shields.io/github/license/oanhnn/laravel-logzio.svg)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/oanhnn/laravel-logzio/master.svg)](https://travis-ci.org/oanhnn/laravel-logzio)
[![Build Status](https://github.com/oanhnn/laravel-logzio/workflows/CI/badge.svg)](https://github.com/oanhnn/laravel-logzio/actions)
[![Coverage Status](https://img.shields.io/coveralls/github/oanhnn/laravel-logzio/master.svg)](https://coveralls.io/github/oanhnn/laravel-logzio?branch=master)
[![Total Downloads](https://img.shields.io/packagist/dt/oanhnn/laravel-logzio.svg)](https://packagist.org/packages/oanhnn/laravel-logzio)
[![Requires PHP](https://img.shields.io/travis/php-v/oanhnn/laravel-logzio.svg)](https://travis-ci.org/oanhnn/laravel-logzio)

Easy integrate [Logz.io](https://logz.io) into PHP and [Laravel](https://laravel.com) 5.6+ Application

## Main features

- [x] Make Logz.io handler for [Monolog](https://packagist.org/packages/monolog/monolog)
- [x] Make `logzio` driver for integrate Logz.io into Laravel Application

## Requirements

* php >=7.1.3
* Laravel 5.6+ (when using with Laravel)

> Laravel 6.0+ required php 7.2+

## Installation

Begin by pulling in the package through Composer.

```bash
$ composer require oanhnn/laravel-logzio
```

## Usage

### PHP (non Laravel)

```php
<?php

use Laravel\Logzio\Log\Handler;
use Monolog\Logger;

$config = [
    'token' => '...',
    'type' => 'http-bulk',
    'ssl' => true,
    'region' => '',
];

$logger = new Logger('log-name');
$logger->pushHandler(new Handler(Logger::DEBUG, true, $config);

$logger->info('Some message');
```

### Laravel

In `config/logging.php` file, config you log with driver `logzio`

```php
<?php
return [
    // ...
	'custom' => [
	    'driver' => 'logzio',
	    'name'   => 'channel-name',
	    'token'  => 'logz-access-token',
	    'type'   => 'http-bulk',
	    'ssl'    => true,
	    'level'  => 'info',
	    'bubble' => true,
	    'region' => 'eu', // leave empty for default region
	],
	// ...
];

```

In your code using

```php
Log::channel('custom')->info('Some message');

```

See more in [Laravel document](https://laravel.com/docs/5.6/logging)

## Changelog

See all change logs in [CHANGELOG](CHANGELOG.md)

## Testing

```bash
$ git clone git@github.com/oanhnn/laravel-logzio.git /path
$ cd /path
$ composer install
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email to [Oanh Nguyen](mailto:oanhnn.bk@gmail.com) instead of 
using the issue tracker.

## Credits

- [Oanh Nguyen](https://github.com/oanhnn)
- [All Contributors](../../contributors)

## License

This project is released under the MIT License.   
Copyright © 2020 [Oanh Nguyen](https://oanhnn.github.io/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).



## [Unreleased] - YYYY-MM-DD

### Added
- Nothing

### Changed
- Nothing

### Deprecated
- Nothing

### Removed
- Nothing

### Fixed
- Nothing

### Security
- Nothing





## [1.1.0] - 2020-09-07

### Added
- Add support Laravel 6.x, 7.x and 8.x

### Changed
- TravisCI to GitHub Workflow





## [1.0.2] - 2019-11-06

### Added
- Support `monolog/monolog:^2.0`

### Changed
- Change coding style from PSR-2 to PSR-12

### Fixed
- Logs wasn't send to Logz.io




## [1.0.1] - 2019-11-06

### Fixed
- Fix class name `ServiceProvider` in `composer.json`. Thank @cschalenborgh




## [1.0.0] - 2019-11-05

### Added
- Add support PHP application (non Laravel)
- Add support `region`, thank @cschalenborgh

### Changed
- Using `GuzzleHttp/Guzzle` package instead of `ext-curl`
- Rename class `Laravel\Logzio\LogzioFormatter` to `Laravel\Logzio\Log\Formatter`
- Rename class `Laravel\Logzio\LogzioHandler` to `Laravel\Logzio\Log\Handler`
- Refactor test scripts




## [0.2.0] - 2019-09-07

### Changed
- Update support for Laravel Framework v6.0
- Update TravisCI configuration
- Update PHPUnit configuration




## [0.1.2] - 2018-07-18

### Fixed
- Missing class `Laravel\Logzio\LogzioServiceProvider`




## [0.1.1] - 2018-05-18

### Changed
- Changed CHANGELOG format

### Fixed
- Fix project link




## 0.1.0 - 2018-04-17

First release



[Unreleased]: https://github.com/oanhnn/laravel-logzio/compare/v1.1.0...develop
[1.1.0]:      https://github.com/oanhnn/laravel-logzio/compare/v1.0.2...v1.1.0
[1.0.2]:      https://github.com/oanhnn/laravel-logzio/compare/v1.0.1...v1.0.2
[1.0.1]:      https://github.com/oanhnn/laravel-logzio/compare/v1.0.0...v1.0.1
[1.0.0]:      https://github.com/oanhnn/laravel-logzio/compare/v0.2.0...v1.0.0
[0.2.0]:      https://github.com/oanhnn/laravel-logzio/compare/v0.1.2...v0.2.0
[0.1.2]:      https://github.com/oanhnn/laravel-logzio/compare/v0.1.1...v0.1.2
[0.1.1]:      https://github.com/oanhnn/laravel-logzio/compare/v0.1.0...v0.1.1


tests/TestCase.php
<?php

namespace Tests;

use Laravel\Logzio\ServiceProvider;
use Orchestra\Testbench\TestCase as Testbench;

/**
 * Class TestCase
 *
 * @package     Tests
 * @author      Oanh Nguyen <oanhnn.bk@gmail.com>
 * @license     The MIT license
 */
class TestCase extends Testbench
{
    /**
     * Define environment setup.
     *
     * @param  \Illuminate\Foundation\Application $app
     * @return void
     */
    protected function getEnvironmentSetUp($app)
    {
        // Define your environment setup.
    }

    /**
     * Get loading package service provider.
     *
     * @param  \Illuminate\Foundation\Application $app
     * @return array
     */
    public function getPackageProviders($app)
    {
        return [
            ServiceProvider::class,
        ];
    }
}


tests/bootstrap.php
<?php

// Set timezone
date_default_timezone_set('UTC');

// Enable Composer autoloader
/** @var \Composer\Autoload\ClassLoader $autoloader */
$autoloader = require dirname(__DIR__) . '/vendor/autoload.php';

// Register test classes
$autoloader->addPsr4('Tests\\', __DIR__);


tests/Unit/LogzioHandlerTest.php
<?php

namespace Tests\Unit;

use GuzzleHttp\Client;
use Laravel\Logzio\Log\Formatter;
use Laravel\Logzio\Log\Handler;
use LogicException;
use PHPUnit\Framework\TestCase;
use Tests\Concerns\CanProvideLogRecord;
use Tests\Concerns\NonPublicAccessible;

/**
 * Class LogzioHandlerTest
 *
 * @package     Tests\Unit
 * @author      Oanh Nguyen <oanhnn.bk@gmail.com>
 * @license     The MIT license
 */
class LogzioHandlerTest extends TestCase
{
    use CanProvideLogRecord;
    use NonPublicAccessible;

    /**
     * Test it should throw exception when create handler without token parameter
     *
     * @return void
     */
    public function testItShouldThowExceptionWhenMissingTokenParameter()
    {
        $this->expectException(LogicException::class);
        $this->expectExceptionMessage('The token parameter is required to use the Logz.io Handler');

        new Handler('debug', true, ['ssl' => true]);
    }

    /**
     * Test it should build endpoint when create handler
     *
     * @param  string $expected
     * @param  array  $options
     * @return void
     * @dataProvider optionsForBuildEndpoint
     */
    public function testItShouldBuildEndpoint($expected, $options)
    {
        $handler = new Handler('debug', true, $options);
        $endpoint = $this->getNonPublicProperty($handler, 'endpoint');

        $this->assertTrue(is_string($endpoint));
        $this->assertSame($expected, $endpoint);
    }

    /**
     * Test it should providers default formatter
     *
     * @return void
     */
    public function testItShouldProvidersDefaultFormatter()
    {
        $handler = new Handler('debug', false, ['token' => 'abc']);
        $defaultFormatter = $handler->getFormatter();

        $this->assertInstanceOf(Formatter::class, $defaultFormatter);

        // Test default formatter be created with default parameters
        $this->assertTrue($defaultFormatter->isAppendingNewlines());
        $this->assertEquals(Formatter::BATCH_MODE_NEWLINES, $defaultFormatter->getBatchMode());
    }

    /**
     * Test it should handle a log record
     *
     * @return void
     */
    public function testItShouldHandleARecord()
    {
        $client = $this->createMock(Client::class);
        $client->expects($this->once())->method('send');

        $handler = new Handler('debug', false, ['token' => 'abc', 'http_client' => $client]);
        $handler->handle($this->logRecord());
    }

    /**
     * Test it should handle a log record
     *
     * @return void
     */
    public function testItShouldHandleBatch()
    {
        $client = $this->createMock(Client::class);
        $client->expects($this->once())->method('send');

        $handler = new Handler('notice', false, ['token' => 'abc', 'http_client' => $client]);
        $handler->handleBatch([
            $this->logRecord('DEBUG', 'test 01'),
            $this->logRecord('INFO', 'test 02'),
            $this->logRecord('alert', 'test 03'),
            $this->logRecord('warning', 'test 04'),
            $this->logRecord('error', 'test 05'),
            $this->logRecord('critical', 'test 06'),
        ]);
    }

    /**
     * The options for create handler and expected endpoint
     *
     * @return array
     */
    public function optionsForBuildEndpoint(): array
    {
        return [
            ['https://listener.logz.io:8071?token=abc', ['token' => 'abc']],
            ['https://listener.logz.io:8071?token=abc', ['token' => 'abc', 'ssl' => true]],
            ['http://listener.logz.io:8070?token=abc', ['token' => 'abc', 'ssl' => false]],
            ['https://listener-au.logz.io:8071?token=abc', ['token' => 'abc', 'region' => 'au']],
            ['https://listener.logz.io:8071?token=abc', ['token' => 'abc', 'region' => 'tw']],
            [
                'https://listener.logz.io:8071?token=abc&type=http-bulk',
                ['token' => 'abc', 'type' => 'http-bulk'],
            ],
            [
                'https://listener-eu.logz.io:8071?token=abc&type=http-bulk',
                ['token' => 'abc', 'type' => 'http-bulk', 'region' => 'eu'],
            ],
        ];
    }
}


src/Log/Handler.php
<?php

namespace Laravel\Logzio\Log;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Psr7\Request;
use LogicException;
use Monolog\Formatter\FormatterInterface;
use Monolog\Handler\AbstractProcessingHandler;
use Monolog\Logger;

/**
 * Class Handler
 *
 * @package     Laravel\Logzio\Log
 * @author      Oanh Nguyen <oanhnn.bk@gmail.com>
 * @license     The MIT license
 *
 * @see    https://support.logz.io/hc/en-us/categories/201158705-Log-Shipping
 * @see    https://app.logz.io/#/dashboard/data-sources/Bulk-HTTPS
 */
final class Handler extends AbstractProcessingHandler
{
    /**
     * The HTTP client
     *
     * @var \GuzzleHttp\ClientInterface
     */
    private $client;

    /**
     * @var string
     */
    private $endpoint;

    /**
     * @param  int|string $level   The minimum logging level to trigger this handler.
     * @param  bool       $bubble  Whether or not messages that are handled should bubble up the stack.
     * @param  array      $options Logz.io client options
     * @throws \LogicException If curl extension is not available.
     */
    public function __construct($level = Logger::DEBUG, bool $bubble = true, array $options = [])
    {
        if (empty($options['token'])) {
            throw new LogicException('The token parameter is required to use the Logz.io Handler');
        }

        $this->client = $this->buildHttpClient($options);
        $this->endpoint = $this->buildEndpoint($options);

        parent::__construct($level, $bubble);
    }

    /**
     * {@inheritdoc}
     */
    public function handleBatch(array $records): void
    {
        $processed = [];

        foreach ($records as $record) {
            if ($this->isHandling($record)) {
                $processed[] = $this->processRecord($record);
            }
        }

        if (!empty($processed)) {
            $this->send($this->getFormatter()->formatBatch($processed));
        }
    }

    /**
     * Build listener endpoint
     *
     * @param  array $options
     * @return string
     */
    protected function buildEndpoint(array $options = []): string
    {
        $region = $options['region'] ?? '';
        $useSsl = $options['ssl'] ?? true;

        $endpoint = sprintf(
            '%s://listener%s.logz.io:%d',
            $useSsl ? 'https' : 'http',
            $this->validRegion($region) ? "-$region" : '',
            $useSsl ? 8071 : 8070
        );

        $endpoint .= '?' . http_build_query(array_filter([
            'token' => $options['token'] ?? null,
            'type' => $options['type'] ?? null,
        ]));

        return $endpoint;
    }

    /**
     * Build HTTP client
     *
     * @param  array $options
     * @return ClientInterface
     */
    protected function buildHttpClient(array $options): ClientInterface
    {
        if (isset($options['http_client']) && $options['http_client'] instanceof ClientInterface) {
            return $options['http_client'];
        }

        return new Client([
            'headers' => [
                'User-Agent' => 'laravel-logzio',
            ],
        ]);
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultFormatter(): FormatterInterface
    {
        return new Formatter();
    }

    /**
     * Send logging data to server
     *
     * @param mixed $data
     * @return void
     */
    protected function send($data)
    {
        $headers = ['Content-Type: application/json'];
        $request = new Request('POST', $this->endpoint, $headers, $data);

        $this->client->send($request);
    }

    /**
     * Validate region
     *
     * @param  string $region
     * @return bool
     */
    protected function validRegion(string $region): bool
    {
        return in_array($region, [
            'au', // Asia Pacific (Sydney) AWS
            'ca', // Canada (Central) AWS
            'eu', // Europe (Frankfurt) AWS
            'nl', // West Europe (Netherlands) Azure
            'wa', // West US 2 (Washington) Azure
        ]);
    }

    /**
     * {@inheritdoc}
     */
    protected function write(array $record): void
    {
        $this->send($record['formatted']);
    }
}


src/Log/Formatter.php
<?php

namespace Laravel\Logzio\Log;

use DateTimeInterface;
use Monolog\Formatter\FormatterInterface;
use Monolog\Formatter\JsonFormatter;

/**
 * Class Formatter
 *
 * @package     Laravel\Logzio\Log
 * @author      Oanh Nguyen <oanhnn.bk@gmail.com>
 * @license     The MIT license
 */
class Formatter extends JsonFormatter implements FormatterInterface
{
    /**
     * Datetime format for Logz.io
     * @see https://support.logz.io/hc/en-us/articles/210206885
     */
    protected const DATETIME_FORMAT = 'c';

    /**
     * @param  int  $batchMode
     * @param  bool $appendNewline
     */
    public function __construct(int $batchMode = self::BATCH_MODE_NEWLINES, bool $appendNewline = true)
    {
        parent::__construct($batchMode, $appendNewline);
    }

    /**
     * Appends the '@timestamp' parameter for Logz.io.
     *
     * @param  array  $record
     * @return string
     */
    public function format(array $record): string
    {
        if (isset($record["datetime"]) && ($record["datetime"] instanceof DateTimeInterface)) {
            $record["@timestamp"] = $record["datetime"]->format(self::DATETIME_FORMAT);

            unset($record["datetime"]);
        }

        return parent::format($record);
    }
}


src/LogzioFormatter.php
<?php

namespace Laravel\Logzio;

use Monolog\Formatter\JsonFormatter;

class LogzioFormatter extends JsonFormatter
{
    /**
     * Datetime format for Logz.io
     * @see https://support.logz.io/hc/en-us/articles/210206885-How-can-I-get-Logz-io-to-read-the-timestamp-within-a-JSON-log-
     */
    const DATETIME_FORMAT = 'c';

    /**
     * @param int $batchMode
     * @param bool $appendNewline
     */
    public function __construct(int $batchMode = self::BATCH_MODE_NEWLINES, bool $appendNewline = true)
    {
        parent::__construct($batchMode, $appendNewline);
    }

    /**
     * Appends the '@timestamp' parameter for Logz.io.
     *
     * @param array $record
     * @return string
     */
    public function format(array $record): string
    {
        if (isset($record["datetime"]) && ($record["datetime"] instanceof \DateTimeInterface)) {
            $record["@timestamp"] = $record["datetime"]->format(self::DATETIME_FORMAT);
            unset($record["datetime"]);
        }

        return parent::format($record);
    }
}


src/ServiceProvider.php
<?php

namespace Laravel\Logzio;

use Illuminate\Support\Facades\Log;
use Illuminate\Support\ServiceProvider as IlluminateServiceProvider;
use Laravel\Logzio\Log\Handler;
use Monolog\Logger;

/**
 * Class Service Provider
 *
 * @package     Laravel\Logzio
 * @author      Oanh Nguyen <oanhnn.bk@gmail.com>
 * @license     The MIT license
 */
class ServiceProvider extends IlluminateServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;

    /**
     * Bootstrap the application services.
     *
     * @return void
     */
    public function boot()
    {
        Log::extend('logzio', function ($app, array $config) {
            // $config = [
            //     'level' => Logger::WARNING,
            //     'bubble' => true,
            //     'token' => '...',
            //     'type' => 'http-bulk',
            //     'ssl' => true,
            //     'region' => '',
            // ];
            $handler = new Handler(
                $config['level'] ?? 'warning',
                $config['bubble'] ?? true,
                $config
            );

            return new Logger($config['name'] ?? $app->environment(), [$handler]);
        });
    }
}


src/LogzioServiceProvider.php
<?php

namespace Laravel\Logzio;

use Illuminate\Support\Facades\Log;
use Illuminate\Support\ServiceProvider;
use Monolog\Logger;

class LogzioServiceProvider extends ServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;

    /**
     * Bootstrap the application services.
     *
     * @return void
     */
    public function boot()
    {
        Log::extend('logzio', function ($app, array $config) {
            $handler = new LogzioHandler(
                $config['token'],
                $config['type'] ?? 'http-bulk',
                $config['ssl'] ?? true,
                $config['level'] ?? Logger::WARNING,
                $config['bubble'] ?? true,
                $config['region'] ?? '' // https://docs.logz.io/user-guide/accounts/account-region.html
            );
            return new Logger($config['name'], [$handler]);
        });
    }
}


src/LogzioHandler.php
<?php

namespace Laravel\Logzio;

use LogicException;
use Monolog\Formatter\FormatterInterface;
use Monolog\Handler\AbstractProcessingHandler;
use Monolog\Handler\Curl\Util;
use Monolog\Logger;

/**
 * @see    https://support.logz.io/hc/en-us/categories/201158705-Log-Shipping
 * @see    https://app.logz.io/#/dashboard/data-sources/Bulk-HTTPS
 */
final class LogzioHandler extends AbstractProcessingHandler
{
    /**
     * @var string
     */
    private $token;

    /**
     * @var string
     */
    private $type;

    /**
     * @var string
     */
    private $endpoint;

    /**
     * @param string $token Log token supplied by Logz.io.
     * @param string $type Host name supplied by Logz.io.
     * @param bool $useSSL Whether or not SSL encryption should be used.
     * @param int|string $level The minimum logging level to trigger this handler.
     * @param bool $bubble Whether or not messages that are handled should bubble up the stack.
     * @param string $region Region code provided by Logz.io (see https://docs.logz.io/user-guide/accounts/account-region.html).
     * @throws \LogicException If curl extension is not available.
     */
    public function __construct(
        string $token,
        string $type = 'http-bulk',
        bool $useSSL = true,
        $level = Logger::DEBUG,
        bool $bubble = true,
        string $region = ''
    ) {
        if (!extension_loaded('curl')) {
            throw new LogicException('The curl extension is needed to use the LogzIoHandler');
        }

        $listener_domain = strlen($region) > 0 ? 'listener-' .$region. '.logz.io' : 'listener.logz.io';

        $this->token = $token;
        $this->type = $type;
        $this->endpoint = $useSSL
            ? 'https://' .$listener_domain. ':8071/'
            : 'http://' .$listener_domain. ':8070/';

        $this->endpoint .= '?' . http_build_query([
            'token' => $this->token,
            'type' => $this->type,
        ]);

        parent::__construct($level, $bubble);
    }

    /**
     * {@inheritdoc}
     */
    public function handleBatch(array $records)
    {
        $level = $this->level;
        $records = array_filter(
            $records,
            function (array $record) use ($level): bool {
                return ($record['level'] >= $level);
            }
        );
        if ($records) {
            $this->send(
                $this->getFormatter()
                    ->formatBatch($records)
            );
        }
    }

    /**
     * Send logging data to server
     *
     * @param mixed $data
     * @return void
     */
    protected function send($data)
    {
        $headers = ['Content-Type: application/json'];

        $ch = curl_init();

        curl_setopt($ch, CURLOPT_URL, $this->endpoint);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

        Util::execute($ch, 3);
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultFormatter(): FormatterInterface
    {
        return new LogzioFormatter();
    }

    /**
     * {@inheritdoc}
     */
    protected function write(array $record)
    {
        $this->send($record['formatted']);
    }
}
